tools/compare.go

```diff
@@ -17,64 +17,137 @@ package tools
+	Short: "Checks for any differences between merge files of two different stores. (To compare the output of two different instrumentation's, for example)",
+	Long: cli.Dedent(`
+		compare-blocks takes in two stores of merged blocks and a range specifying the first x blocks you want to compare.
+		It operates in chunks of 100,000 blocks. Chunks that contain a difference will be communicated as well as the blocks within 
+		that contain differences. Chunks that do not have any differences will be outputted as identical.
+		
+		After passing through the chunks, it will output instructions on how to locate a specific difference based on the
+		blocks that were given. This is done by applying the --diff=true flag before your args. 
+
+		Commands inputted with --diff=true will display the blocks that have differences, as well as the difference. 
+		These commands should have a range that matches 1 bundle, so a range of 100.
+	`),
+	Args: cobra.ExactArgs(3),
+	RunE: compareBlocksE,
+	Example: cli.Dedent(`
+		# Run over full block range
+		fireeth tools compare-blocks sf_bundle/ cs_bundle/ 100-200
+
+		# Run over specific block range (inclusive/inclusive)
+		fireeth tools compare-blocks --diff=true sf_bundle/ cs_bundle/ 100-200
+	`),
 }
 
 func init() {
 	Cmd.AddCommand(compareBlocksCmd)
+	compareBlocksCmd.PersistentFlags().Bool("diff", false, "A flag to check for differences over small range")
 }
 
 func compareBlocksE(cmd *cobra.Command, args []string) error {
+	fmt.Printf("\n-----starting comparison-----\n")
 	ctx := cmd.Context()
 
-	storeADef := args[0]
-	storeBDef := args[1]
+	storeExpectedDef := args[0]
+	storeReceivedDef := args[1]
+	blockRange := block.ParseRange(args[2])
 
-	blockNum, err := strconv.ParseUint(args[2], 10, 64)
+	storeExpected, err := dstore.NewDBinStore(storeExpectedDef)
 	if err != nil {
-		return fmt.Errorf("unable to parse block number %q: %w", args[0], err)
+		return fmt.Errorf("unable to create store at path %q: %w", storeExpectedDef, err)
 	}
 
-	storeA, err := dstore.NewDBinStore(storeADef)
+	storeReceived, err := dstore.NewDBinStore(storeReceivedDef)
 	if err != nil {
-		return fmt.Errorf("unable to create store at path %q: %w", storeADef, err)
+		return fmt.Errorf("unable to create store at path %q: %w", storeReceivedDef, err)
 	}
 
-	storeB, err := dstore.NewDBinStore(storeBDef)
-	if err != nil {
-		return fmt.Errorf("unable to create store at path %q: %w", storeBDef, err)
-	}
+	blocksExpected := make(map[uint64]*pbeth.Block)
+	blocksReceived := make(map[uint64]*pbeth.Block)
+
+	collectBlocksExpected := func(store dstore.Store, blockMap map[uint64]*pbeth.Block) error {
+		fmt.Printf("collecting expected blocks\n")
+		var files []string
+		err = storeExpected.Walk(ctx, "", func(filename string) (err error) {
+			files = append(files, filename)
+			return nil
+		})
+
+		var toClose []io.ReadCloser
+		defer func() {
+			for i := range toClose {
+				err := toClose[i].Close()
+				if err != nil {
+					return
+				}
```

Comment:
That is quite an improper error handling. Imagine there is N elements in `toClose` array, if `err != nil` on first element in `toClose`, `return` is called and no other handle will be close, leaving dangling file handles. Also, the actual error message is lost because `err` is not used anywhere, so you stop in between and error is never visible.

In general also `.Close` error can be something a bit trickier to handle because they are often done in a `defer`, so `return` exit the defer handled but does not impact the outer function in which `defer` is used. In this specific case, I would do a panic if we can't close the file for now.

But code will be refactored, so the exact way might change specifically if `Close` is closer to its usage.

Reply:
This will require an heavy refactor of how the command works. 

While you walk `expected` store, for each received bundle, right away you should try to fetch the bundle on the `actual` store. If it's not in there, you exit with an error right away.

Then, you have two bundle, you compare them right away. To limit memory, you should also keep in memory the offending bundle name and nothing else for the summary. When actually walking the summary offending bundle list, then you compute on the fly the proper range.

We can discuss all this tomorrow.